### PR TITLE
Password policy validation fail for new users

### PR DESCRIPTION
**Pull request by MoonISO**
Password policy validation fail for new users #22014

1. Create a new GenerateSecurePasswordEvent
2. Dispatch the event
3. If the passwordEvent is null use fallback

Other option for fallback:
`/apps/provisioning_api/lib/Controller/UsersController.php` (around line324)
```
...
$passwordEvent = new GenerateSecurePasswordEvent();
$this->eventDispatcher->dispatchTyped($passwordEvent);
$password .= $passwordEvent->getPassword() ?? $this->secureRandom->generate(21);
...
```

### CODE REVIEW

File: apps/provisioning_api/lib/Controller/UsersController.php [hunk 1]
```diff
@@ -286,9 +312,19 @@ public function addUser(string $userid,
 				throw new OCSException('To send a password link to the user an email address is required.', 108);
 			}
 
-			$password = $this->secureRandom->generate(10);
-			// Make sure we pass the password_policy
-			$password .= $this->secureRandom->generate(2, '$!.,;:-~+*[]{}()');
+ 			$passwordEvent = new GenerateSecurePasswordEvent();
+			$this->eventDispatcher->dispatchTyped($passwordEvent);
+			$passwordEvent->getPassword();
```
Comment (kesselb) [inline apps/provisioning_api/lib/Controller/UsersController.php:317]:
```suggestion
			$password = $passwordEvent->getPassword();
```

File: apps/provisioning_api/lib/Controller/UsersController.php [hunk 2]
```diff
@@ -286,9 +312,19 @@ public function addUser(string $userid,
...
+			if(isset($passwordEvent)){
+				$password .= $passwordEvent->getPassword();
+			} else {
```
Comment (kesselb) [inline apps/provisioning_api/lib/Controller/UsersController.php:320]:
```
if ($password === null) {
  // fallback logic for password
}
```

If $password is null the password_policy (or any other listener to GenerateSecurePasswordEvent) is not enabled hence we can use a simpler fallback but using a strong password by default is probably not wrong.

**Review by kesselb — COMMENTED** [review]
Thanks :+1:

**Comment by MoonISO** [comment]
I added your suggested modifications. Unfortunately, I am messing a bit up with PHP style recommendations.

**Comment by kesselb** [comment]
Hi @MoonISO,

thanks again for your pull request :+1: I'm not able to push changes to your branch so I redid the changes and updated the tests: #22145.

**Comment by MorrisJobke** [comment]
> thanks again for your pull request 👍 I'm not able to push changes to your branch so I redid the changes and updated the tests: #22145.

I guess it should be #22159

**Pull request closed without merging**